src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1126,6 +1126,174 @@ object Desugar {
+              (dTerPre, dTer) <- prelude(option(spec.terminationMeasure map terminationMeasureD(ctx)))
+
+              dBody = blockD(ctx)(body)
+
+              continueLabelName = nm.continueLabel(n, info)
+              continueLoopLabelProxy = in.LabelProxy(continueLabelName)(src)
+              continueLoopLabel = in.Label(continueLoopLabelProxy)(src)
+
+              breakLabelName = nm.breakLabel(n, info)
+              breakLoopLabelProxy = in.LabelProxy(breakLabelName)(src)
+              _ <- declare(breakLoopLabelProxy)
+              breakLoopLabel = in.Label(breakLoopLabelProxy)(src)
+
+              wh = if (shorts.length == 2) {
+                // here we know we have the value variable which we create
+                val valueSrc = meta(shorts(1))
+                val valueLeft = assignableVarD(ctx)(shorts(1))
+                val valueVar = in.Assignee.Var(valueLeft)
+                val valueAss = singleAss(valueVar, in.IndexedExp(copiedVar, in.IntLit(0)(valueSrc), typ)(valueSrc))(valueSrc)
+                val updateValue = in.If(cond, singleAss(valueVar, in.IndexedExp(copiedVar, indexLeft, typ)(valueSrc))(valueSrc), in.Seqn(Vector())(valueSrc))(valueSrc)
+
+                // we also need an invariant that states that
+                // for index i and value j and range expression x
+                // invariant 0 <= i && i < len(x) ==> j == x[i]
+                val indexValueEq = in.Implication(
+                  in.And(
+                    in.AtMostCmp(in.IntLit(0)(indexSrc), indexLeft)(indexSrc),
+                    in.LessCmp(indexLeft, length)(indexSrc))(indexSrc),
+                  in.ExprAssertion(in.EqCmp(in.IndexedExp(copiedVar, indexLeft, typ)(rangeExpSrc), valueLeft)(rangeExpSrc))(rangeExpSrc)
+              )(rangeSrc)
+                in.Block(
+                  Vector(valueLeft.asInstanceOf[in.LocalVar]),
+                  Vector(copyAss, indexAss, valueAss) ++ dInvPre ++ dTerPre ++ Vector(
+                    in.While(cond, dInv ++ Vector(indexBoundsInv, indexValueEq), dTer, in.Block(Vector(continueLoopLabelProxy),
+                      Vector(dBody, continueLoopLabel, incrIndex, updateValue) ++ dInvPre ++ dTerPre
+                    )(src))(src), breakLoopLabel
+                  )
+                )(src)
+              }
+              else {
+                // else we do not have a value variable and the while loop has only
+                // the index in bounds invariant added
+                in.Seqn(
+                  Vector(copyAss, indexAss) ++ dInvPre ++ dTerPre ++ Vector(
+                    in.While(cond, dInv ++ Vector(indexBoundsInv), dTer, in.Block(Vector(continueLoopLabelProxy),
+                      Vector(dBody, continueLoopLabel, incrIndex) ++ dInvPre ++ dTerPre
+                    )(src))(src), breakLoopLabel
+                  )
+                )(src)
+              }
+            } yield wh))
+
+          case n@PAssForRange(range, ass, spec, body) =>
```

Comment:
a similar comment as before applies here - how much of this logic is repeated? would it make sense to refactor this to avoid repetition?

Reply:
The logic is the same, but now instead of a vector of PIdnUnk objects that I have to create variables from I have a bunch of lvalues so the way they are handled is different. I am not aware of a smart way to handle them identically.